### PR TITLE
[FIX] product,sale(_*): pricelists multi-company checks

### PR DESCRIPTION
**Pull request by mathysPaul**
Since 83c52575d0ce3d7deb1737c622c6ce366db8c31b, there
is no default pricelist in new database, and a magic
pricelist is created for each company when the pricelist
setting is enabled.

But, since this automatically created pricelist is restricted
to a specific company, you couldn't use it as 'Base' for
rules of other pricelists, which are, by default, shared
between companies, unless you set a specific company on them.

This commit changes that default behavior, to restrict
pricelists to the current company (unless the field is
manually emptied), to avoid those multi-company issues,
particularly in single-company databases where
multi-company issues are not expected.

Also, while testing this task, we discovered that a user
could change the company of a pricelist, even if it conflicted
with existing rules.  To avoid this situation, we now trigger
the multi-company check on the pricelist rules when the
pricelist company is updated.

task-3610823

See odoo/enterprise#55857

### CODE REVIEW

File: addons/product/models/product_pricelist_item.py [hunk 1]
```diff
@@ -460,3 +460,9 @@ def _compute_price_before_discount(self, *args, **kwargs):
             pricelist_rule = pricelist_item
 
         return pricelist_rule._compute_base_price(*args, **kwargs)
+
+    @api.constrains('base_pricelist_id')
+    def _check_websites_in_company(self):
+        for record in self:
+            if record.base_pricelist_id.company_id != record.pricelist_id.company_id:
```
Comment (Demesmaeker) [inline addons/product/models/product_pricelist_item.py:467]: While it's true that if there is a pricelist on the base_pricelist, then the new one (based
 on it), must have the same one, the opposite is false.
If the base pricelist work with every company, why couldn't the new pricelist be restricted to a company?

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello :duck:

File: addons/product/models/product_pricelist_item.py [hunk 2]
```diff
@@ -460,3 +460,9 @@ def _compute_price_before_discount(self, *args, **kwargs):
...
+                raise ValidationError(_("You can not link this price list to another price list if the company restrictions are not the same."))
```
[the remark this answers is not in the document]
Comment (dalaOdoo) [reply addons/product/models/product_pricelist_item.py:468]: yes it's ok but the funny part is missing :-)

File: addons/sale/tests/test_sale_prices.py [hunk 3]
```diff
@@ -408,6 +408,7 @@ def test_multi_currency_discount(self):
 
         pricelist = self.env['product.pricelist'].create({
             'name': 'Test multi-currency',
+            'company_id': self.env['res.company'],
```
Comment (Demesmaeker) [inline addons/sale/tests/test_sale_prices.py:411]: default

File: addons/sale/tests/common.py [hunk 4]
```diff
@@ -93,6 +93,7 @@ def setup_sale_configuration_for_company(cls, company):
             'default_pricelist': cls.env['product.pricelist'].with_company(company).create({
                 'name': 'default_pricelist',
                 'currency_id': company.currency_id.id,
+                'company_id': cls.env['res.company'],
```
Comment (Demesmaeker) [inline addons/sale/tests/common.py:96]: default

File: addons/product/tests/test_product_pricelist.py [hunk 5]
```diff
@@ -69,12 +69,6 @@ def setUpClass(cls):
         cls.customer_pricelist = cls.env['product.pricelist'].create({
             'name': 'Customer Pricelist',
             'item_ids': [
-                Command.create({
-                    'name': 'Default pricelist',
-                    'compute_price': 'formula',
-                    'base': 'pricelist',
-                    'base_pricelist_id': cls.pricelist.id,
-                }),
```
Comment (Demesmaeker) [inline addons/product/tests/test_product_pricelist.py:77]: why this change?

File: addons/sale/tests/test_sale_report.py [hunk 6]
```diff
@@ -43,9 +43,9 @@ def test_sale_report_foreign_currency(self):
 
         # Create corresponding pricelists and rates.
         pricelists = self.env['product.pricelist'].create([
-            {'name': 'Pricelist (USD)', 'currency_id': usd.id},
-            {'name': 'Pricelist (EUR)', 'currency_id': eur.id},
-            {'name': 'Pricelist (ARS)', 'currency_id': ars.id},
+            {'name': 'Pricelist (USD)', 'currency_id': usd.id, 'company_id': self.env['res.company']},
+            {'name': 'Pricelist (EUR)', 'currency_id': eur.id, 'company_id': self.env['res.company']},
+            {'name': 'Pricelist (ARS)', 'currency_id': ars.id, 'company_id': self.env['res.company']},
```
Comment (Demesmaeker) [inline addons/sale/tests/test_sale_report.py:48]: Quite the opposite in fact, in the follow-up of this test, we are then using these pricelists on 2 different companies... As the default is the env company, you should instead voluntarily enforce that this value stays empty. It's strange that the SO can be made with a pricelist belonging to another company, in fact. I don't see why this test isn't currently failing... :thinking:

File: addons/product/models/product_pricelist_item.py [hunk 7]
```diff
@@ -460,3 +460,11 @@ def _compute_price_before_discount(self, *args, **kwargs):
...
 
         return pricelist_rule._compute_base_price(*args, **kwargs)
+
+    @api.constrains('base_pricelist_id')
+    def _check_company_pricelist_in_multi_company_context(self):
+        for record in self:
+            if (record.base_pricelist_id.company_id
+                and record.base_pricelist_id.company_id != record.pricelist_id.company_id):
+                raise ValidationError(_("You can not link this pricelist to another pricelist if \
+                    the company restrictions are not the same."))
```
Comment (Demesmaeker) [inline addons/product/models/product_pricelist_item.py:470]: To break lines, we usually follow this:

```py
quacking_ducks_messages = ["quack", "quack", "quack"]
quacking_ducks_messages = [
    "quack", "quack", "quack", "quacking intensifies", "QUACK!", "QUACK", "QUAAAACK!"
]
quacking_ducks_messages = [
    "quackquackquackquackquackquack", 
    "quackquackquackquack", 
    "quackquackquack", 
    "quacking intensifies", 
    "QUACK! QUACK!!", 
    "QUAAAACK!",
]
```

File: addons/product/models/product_pricelist_item.py [hunk 8]
```diff
@@ -460,3 +460,11 @@ def _compute_price_before_discount(self, *args, **kwargs):
             pricelist_rule = pricelist_item
 
         return pricelist_rule._compute_base_price(*args, **kwargs)
+
+    @api.constrains('base_pricelist_id')
+    def _check_company_pricelist_in_multi_company_context(self):
+        for record in self:
+            if (record.base_pricelist_id.company_id
+                and record.base_pricelist_id.company_id != record.pricelist_id.company_id):
```
Comment (Demesmaeker) [inline addons/product/models/product_pricelist_item.py:468]: Though for the if statement, this works fine too:
```suggestion
            if record.base_pricelist_id.company_id \
               and record.base_pricelist_id.company_id != record.pricelist_id.company_id:
```

**Review by Demesmaeker — CHANGES_REQUESTED** [review]

File: addons/sale/tests/common.py (hunk 4 above)
Comment (Demesmaeker) [reply addons/sale/tests/common.py:96]: My mistake, it's an empty recordset, I wrongly read `cls.env.company`

File: addons/sale/tests/test_sale_prices.py (hunk 3 above)
Comment (Demesmaeker) [reply addons/sale/tests/test_sale_prices.py:411]: My mistake, it's an empty recordset, I wrongly read `cls.env.company`

File: addons/sale/tests/test_sale_report.py (hunk 6 above)
Comment (Demesmaeker) [reply addons/sale/tests/test_sale_report.py:48]: My mistake, it's an empty recordset, I wrongly read `cls.env.company`

File: addons/product/models/product_pricelist.py [hunk 9]
```diff
@@ -33,7 +33,9 @@ def _default_currency_id(self):
     company_id = fields.Many2one(
         comodel_name='res.company',
         tracking=5,
+        default=lambda self: self.env.company,
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/product/models/product_pricelist.py:36]: This comment was not resolved.

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
edit: nevermind, it'd in fact work as it's called at the record creation, and not on SQL level, and since you now allowed the base pricelist to be wothout company even if the new one isn't, it's okay.

lgtm then :partying_face:

**Review by Demesmaeker — APPROVED** [review]

**Review by mathysPaul — COMMENTED** [review]
@Feyensv Some questions :)

File: addons/product/models/product_pricelist.py [hunk 10]
```diff
@@ -69,6 +71,13 @@ def _compute_display_name(self):
         for pricelist in self:
             pricelist.display_name = f'{pricelist.name} ({pricelist.currency_id.name})'
 
+    def write(self, values):
+        res = super().write(values)
+        if 'company_id' in values and len(self) == 1:
+            self.item_ids._check_company()
+
+        return res
+
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/models/product_pricelist.py:80]: We'll go through those changes tomorrow together ;)

**Comment by Feyensv** [comment]
@robodoo r+
